pkg/apis/resource/types.go

```diff
@@ -0,0 +1,388 @@
+// (AllocationModeWaitForFirstConsumer). Other modes might get added in the
+// future.
+type AllocationMode string
+
+const (
+	// When a ResourceClaim has AllocationModeWaitForFirstConsumer, allocation is
+	// delayed until a Pod gets scheduled that needs the ResourceClaim. The
+	// scheduler will consider all resource requirements of that Pod and
+	// trigger allocation for a node that fits the Pod.
+	AllocationModeWaitForFirstConsumer AllocationMode = "WaitForFirstConsumer"
+
+	// When a ResourceClaim has AllocationModeImmediate, allocation starts
+	// as soon as the ResourceClaim gets created. This is done without
+	// considering the needs of Pods that will use the ResourceClaim
+	// because those Pods are not known yet.
+	AllocationModeImmediate AllocationMode = "Immediate"
+)
+
+// ResourceClaimStatus tracks whether the resource has been allocated and what
+// the resulting attributes are.
+type ResourceClaimStatus struct {
+	// DriverName is a copy of the driver name from the ResourceClass at
+	// the time when allocation started.
+	DriverName string
+
+	// Allocation is set by the resource driver once a resource has been
+	// allocated successfully. If this is not specified, the resource is
+	// not yet allocated.
+	Allocation *AllocationResult
+
+	// ReservedFor indicates which entities are currently allowed to use
+	// the claim. A Pod which references a ResourceClaim which is not
+	// reserved for that Pod will not be started.
+	//
+	// There can be at most 32 such reservations. This may get increased in
+	// the future, but not reduced.
+	ReservedFor []ResourceClaimConsumerReference
+
+	// DeallocationRequested indicates that a ResourceClaim is to be
+	// deallocated.
+	//
+	// The driver then must deallocate this claim and reset the field
+	// together with clearing the Allocation field.
+	//
+	// While DeallocationRequested is set, no new consumers may be added to
+	// ReservedFor.
+	DeallocationRequested bool
+}
+
+// ReservedForMaxSize is the maximum number of entries in
+// claim.status.reservedFor.
+const ResourceClaimReservedForMaxSize = 32
+
+// AllocationResult contains attributed of an allocated resource.
+type AllocationResult struct {
+	// ResourceHandle contains arbitrary data returned by the driver after a
+	// successful allocation. This is opaque for
+	// Kubernetes. Driver documentation may explain to users how to
+	// interpret this data if needed.
+	//
+	// The maximum size of this field is 16KiB. This may get
+	// increased in the future, but not reduced.
+	ResourceHandle string
+
+	// This field will get set by the resource driver after it has
+	// allocated the resource driver to inform the scheduler where it can
+	// schedule Pods using the ResourceClaim.
+	//
+	// Setting this field is optional. If null, the resource is available
+	// everywhere.
+	AvailableOnNodes *core.NodeSelector
+
+	// ShareableResource determines whether the resource supports more
+	// than one user at a time.
+	ShareableResource bool
```

Comment:
I had a numeric limit earlier, but the exact semantic was problematic because it wasn't well-defined what it should count - there could be limits for nodes and/or pods.

We decided to postpone that and only do "one consumer" vs. "unlimited". Additional limits can be added later, with "Shareable: true", "MaxPerNode: 5" or something like that.

Reply:
Still - does shareable mean "on a node" or "in a cluster" ?

Reply:
No limitations whatsoever.